Solana NFT gating feature for web3 premium users

hadiamjad opened the pull request:
Tested locally and working fine.
Known Issues: https://docs.google.com/document/d/1bPsoMVDmXR69tf5TKJsOwQhszCQMYYjaSlX5GH92LJI/edit?usp=sharing

kdenhartog commented on package.json:
```diff
@@ -50,14 +50,12 @@
   },
   "dependencies": {
     "@emotion/react": "11.11.1",
+    "@solana/web3.js": "^1.77.3",
```
can we version lock this here?

kdenhartog commented on src/components/web3/api.ts:
```diff
@@ -221,6 +241,46 @@ export const web3Prove = async (
+    // HT:https://stackoverflow.com/questions/40031688/javascript-arraybuffer-to-hex/40031979
+    // btoa has some characters not allowed by the EIP-4361 ABNF
+    nonce: nonce,
+    issuedAt: new Date().toISOString(),
```
Are we relying upon this time within the backend at all? If so this is potentially modifiable by the client so we'll want to check that verification logic.

TODO for @kdenhartog or @bcaller to look into the impact of malicious times

kdenhartog replied:
Looks like this one is an issue: https://github.com/brave-intl/subscriptions/pull/367#discussion_r1238032632

kdenhartog commented on src/components/web3/api.ts:
```diff
@@ -13,7 +14,7 @@ const SIMPLEHASH_PROXY_ROOT_URL = "/api/v1/simplehash";
   method: string;
   proof: {
-    signer: string;
+    signer: any;
```
nit: What's the value in changing this to an any type? Seems like keeping it as a string would be more useful here.

hadiamjad replied:
fixed it. Back to `string`

kdenhartog commented on src/hooks/use-web3-call-state.ts:
```diff
@@ -187,3 +204,174 @@ export function useWeb3CallState(
   };
 }
+
+export function useWeb3SolCallState(
```
nit: It seems that a lot of this logic is copied from the `useWeb3CallState` function above. What's the reason for reimplementing it on this branching logic rather than branching inside the function we've already declared?

kdenhartog commented on src/components/web3/SolLogin.tsx:
```diff
@@ -0,0 +1,68 @@
+  const { t } = useTranslation();
+
+  useEffect(() => {
+    window.braveSolana?.on("!!! accountChanged", () => setNotice(undefined));
```
same here regarding using `window.solana.*` usage

kdenhartog commented on src/components/web3/SolLogin.tsx:
```diff
@@ -0,0 +1,68 @@
+                  href="https://support.brave.com/hc/en-us/articles/4415497656461-Brave-Wallet-FAQ"
+                  style={{ color: "inherit" }}
+                >
+                  Help with Brave Wallet
```
nit: Are we assuming users will only use brave wallet here to connect?

mrose17 replied:
The requirements that I know of:
- to create a room, you have to be using the Brave browser
- to create a web3 room, you need to have a wallet, though the Brave wallet is preferred ... all the Web3 Brave Talk documentation talks about only the Brave wallet, as a big part of web3 Brave Talk is to promote the  Brave wallet.

kdenhartog replied:
> to create a web3 room, you need to have a wallet, though the Brave wallet is preferred ... all the Web3 Brave Talk documentation talks about only the Brave wallet, as a big part of web3 Brave Talk is to promote the Brave wallet.

My thinking is that users may want to utilize a different wallet if they're joining the call. Ultimately, this is a call by your team/product so non-blocking. I just wanted to call it out in case we were unintentionally causing friction the UX for non-brave wallet users. If it's intentional no concerns from me.

mrose17 replied:
I'm looking at the link -- https://brave.com/web3-talk-calls/ -- and it says ```Web3 calls are available to all Brave Talk users on any browser with any Web3 wallet. ``` which is what's on the web3 Brave Talk homepage.  is there some other/additional text that you'd recommend to that page?

@hadiamjad - i couldn't find the link that Kyle is referring to above. --  ttps://support.brave.com/hc/en-us/articles/4415497656461-Brave-Wallet-FAQ -- can you tell me how to find it? thanks!

hadiamjad replied:
@mrose17  yea when Login fails or user does allow a selected account, the user is promoted with this link. You can find it `Login.tsx` (https://github.com/brave/brave-talk/blob/7aadf34ffe8dc13c2a3e6e85bfa6698d3e500c69/src/components/web3/Login.tsx#L37)and `SolLogin.tsx` (https://github.com/brave/brave-talk/blob/7aadf34ffe8dc13c2a3e6e85bfa6698d3e500c69/src/components/web3/SolLogin.tsx#L40)
<img width="750" alt="image" src="https://github.com/brave/brave-talk/assets/46374292/f0123c4a-527b-4365-8da6-fa64c7913397">

mrose17 replied:
Thanks. Well, that is a natural place to put the text and that link is where they need to go. The only other thing I think we could do is have another line saying "Not using Brave Wallet?" but I have no idea where that link should go. So, i don't think there is anything that can be done.

kdenhartog replied:
I'd think an error message suggesting that user's contact their wallet's support channel would be useful. Otherwise a link to community channel would also be nice so that we can get reports of these failures and I can reach out to the appropriately wallet devs to help us be able to resolve the issue.

This is just a nit though, so that's up to you all to decide and I'm fine resolving with no action.

kdenhartog commented on src/components/web3/StartCallSol.tsx:
A lot of this code seems redundant to the code within `StartCall.tsx`. Can we consider consolidating this code.

mrose17 replied:
Good point, but that's probably post-FCS, for two reasons: first, the models used by ETH and SOL wallet support are somewhat different, so consolidation is likely to be viable for half of the code, but not the other half; second, moving to a unified approach, e.g., the newly-implemented [Wallet standard](https://github.com/wallet-standard/). I'm not opposed to simplifying, but I'd rather not de-stabilize things...

mrose17 replied:
I believe that @hadiamjad is making a pass now to see if there are "obvious" things to be simplified/consolidated.  I'm simply not sure how much can be done in terms of complete consolidation (-;

hadiamjad replied:
I have removed extra branches and redundant code.

kdenhartog replied:
cool, FWIW none of these issues are sec blocking. Just wanted to call it out because it impacted the readability of the code a bit for me.

kdenhartog reviewed (changes requested):
Seems very well aligned with the ETH work. Left a few comments to address at this point

kdenhartog commented:
> **New and updated dependency changes detected.** Learn more about [Socket for GitHub ↗︎](https://socket.dev?utm_medium=gh)
> 
> Packages		Version	New capabilities	Transitives[1](#user-content-fn-1-53311aef44133c123b74d6b3f6414665)	Size	Publisher
> [@solana/web3.js](https://socket.dev/npm/package/@solana/web3.js/overview/1.77.3)	🆕	[1.77.3](https://socket.dev/npm/package/@solana/web3.js/overview/1.77.3)	[eval](https://socket.dev/npm/package/@solana/web3.js/issues/1.77.3?issue=usesEval), [network](https://socket.dev/npm/package/@solana/web3.js/issues/1.77.3?issue=networkAccess), [environment](https://socket.dev/npm/package/@solana/web3.js/issues/1.77.3?issue=envVars)	`+38`	22.9 MB	[steveluscher](https://socket.dev/npm/user/steveluscher)
> ## Footnotes
> 1. [docs.socket.dev](https://docs.socket.dev) [↩](#user-content-fnref-1-53311aef44133c123b74d6b3f6414665)

The size of this dependency is quite massive for a web app which could slow the responsiveness of the App down. I think we could implement this without needing to bringing it in to keep the size down here.

mrose17 commented:
> > **New and updated dependency changes detected.** Learn more about [Socket for GitHub ↗︎](https://socket.dev?utm_medium=gh)
> > Packages		Version	New capabilities	Transitives[1](#user-content-fn-1-53311aef44133c123b74d6b3f6414665)	Size	Publisher
> > [@solana/web3.js](https://socket.dev/npm/package/@solana/web3.js/overview/1.77.3)	🆕	[1.77.3](https://socket.dev/npm/package/@solana/web3.js/overview/1.77.3)	[eval](https://socket.dev/npm/package/@solana/web3.js/issues/1.77.3?issue=usesEval), [network](https://socket.dev/npm/package/@solana/web3.js/issues/1.77.3?issue=networkAccess), [environment](https://socket.dev/npm/package/@solana/web3.js/issues/1.77.3?issue=envVars)	`+38`	22.9 MB	[steveluscher](https://socket.dev/npm/user/steveluscher)
> > ## Footnotes
> > 
> > 1. [docs.socket.dev](https://docs.socket.dev) [↩](#user-content-fnref-1-53311aef44133c123b74d6b3f6414665)
> 
> The size of this dependency is quite massive for a web app which could slow the responsiveness of the App down. I think we could implement this without needing to bringing it in to keep the size down here.

Understood. Once this PR is in place, the next item is https://github.com/brave/brave-talk/issues/833 ... for myself, i have never understood why the ETH and SOL and ... repositories are so monolithic.

mrose17 commented on src/components/web3/api.ts:
```diff
@@ -51,9 +63,17 @@ export const web3Login = async (): Promise<string> => {
 };
 
+export const web3LoginSol = async (): Promise<string> => {
+  const result = await window.braveSolana.connect();
```
@kdenhartog - here is what I don't understand. our documentation says `window.braveSolana.connect` - why have such an object if we also have `window.solana` ???

@hadiamjad - until we have a definitive answer on this, please continue to use `window.braveSolana`, but be prepared to use `window.solana` as a fallback... cf., https://bravesoftware.slack.com/archives/CBP2ZKC0H/p1687430956921859

kdenhartog replied:
Oh, we probably need to update our docs then. I added a comment in that slack message because I realized as soon as you asked that question (thank you for raising it!) I realized directly calling `window.solana` is actually no longer the BCP. The `window.braveSolana` object was added so we could support wallet-standard. We could even simplify this logic a bit by reusing Solana lab's [component library](https://github.com/wallet-standard/wallet-standard/tree/master/packages/react/core).

hadiamjad commented:
> > **New and updated dependency changes detected.** Learn more about [Socket for GitHub ↗︎](https://socket.dev?utm_medium=gh)
> > Packages		Version	New capabilities	Transitives[1](#user-content-fn-1-53311aef44133c123b74d6b3f6414665)	Size	Publisher
> > [@solana/web3.js](https://socket.dev/npm/package/@solana/web3.js/overview/1.77.3)	🆕	[1.77.3](https://socket.dev/npm/package/@solana/web3.js/overview/1.77.3)	[eval](https://socket.dev/npm/package/@solana/web3.js/issues/1.77.3?issue=usesEval), [network](https://socket.dev/npm/package/@solana/web3.js/issues/1.77.3?issue=networkAccess), [environment](https://socket.dev/npm/package/@solana/web3.js/issues/1.77.3?issue=envVars)	`+38`	22.9 MB	[steveluscher](https://socket.dev/npm/user/steveluscher)
> > ## Footnotes
> > 
> > 1. [docs.socket.dev](https://docs.socket.dev) [↩](#user-content-fnref-1-53311aef44133c123b74d6b3f6414665)
> 
> The size of this dependency is quite massive for a web app which could slow the responsiveness of the App down. I think we could implement this without needing to bringing it in to keep the size down here.

I have removed the `web3/solana` dependency to avoid a massive size increase and used different package `buffer`

kdenhartog commented:
> > > **New and updated dependency changes detected.** Learn more about [Socket for GitHub ↗︎](https://socket.dev?utm_medium=gh)
> > > Packages		Version	New capabilities	Transitives[1](#user-content-fn-1-53311aef44133c123b74d6b3f6414665)	Size	Publisher
> > > [@solana/web3.js](https://socket.dev/npm/package/@solana/web3.js/overview/1.77.3)	🆕	[1.77.3](https://socket.dev/npm/package/@solana/web3.js/overview/1.77.3)	[eval](https://socket.dev/npm/package/@solana/web3.js/issues/1.77.3?issue=usesEval), [network](https://socket.dev/npm/package/@solana/web3.js/issues/1.77.3?issue=networkAccess), [environment](https://socket.dev/npm/package/@solana/web3.js/issues/1.77.3?issue=envVars)	`+38`	22.9 MB	[steveluscher](https://socket.dev/npm/user/steveluscher)
> > > ## Footnotes
> > > 
> > > 1. [docs.socket.dev](https://docs.socket.dev) [↩](#user-content-fnref-1-53311aef44133c123b74d6b3f6414665)
> > 
> > 
> > The size of this dependency is quite massive for a web app which could slow the responsiveness of the App down. I think we could implement this without needing to bringing it in to keep the size down here.
> 
> I have removed the `web3/solana` dependency to avoid a massive size increase and used different package `buffer`

Awesome change! Thanks for doing that. No security concerns from me to use `buffer` here either.

kdenhartog commented on src/components/web3/api.ts:
```diff
@@ -221,6 +241,46 @@ export const web3Prove = async (
+  const payloadBytes = new TextEncoder().encode(payload);
+  const hexPayload = hexlify(payloadBytes);
+  const { publicKey, signature } = await window.braveSolana.signMessage(
+    payloadBytes
```
confirmed this isn't the case - please ignore this one.

kdenhartog reviewed (approved):
https://github.com/brave/brave-talk/pull/1022#discussion_r1237993838 is the only sec related concern, but it's best solved in the backend, so approving this.

mrose17 reviewed (approved):
Great work! Very thorough. Good use of the query flag to hide the feature in production, while we iron out some issues with the wallet, etc.
